adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,267 @@
+          end
+
+          def full_http_request_url(rack_request)
+            rack_request.url
+          end
+
+          # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name
+          #
+          # recommendation: span.name(s) should be low-cardinality (e.g.,
+          # strip off query param value, keep param name)
+          #
+          # see http://github.com/open-telemetry/opentelemetry-specification/pull/416/files
+          def create_request_span_name(request_uri_or_path_info)
+            # NOTE: dd-trace-rb has implemented 'quantization' (which lowers url cardinality)
+            #       see Datadog::Quantization::HTTP.url
+
+            if (implementation = config[:url_quantization])
+              implementation.call(request_uri_or_path_info)
+            else
+              request_uri_or_path_info
+            end
+          end
+
+          def base_url(rack_request)
+            if rack_request.respond_to?(:base_url)
+              rack_request.base_url
+            else
+              # Compatibility for older Rack versions
+              rack_request.url.chomp(rack_request.fullpath)
+            end
+          end
+
+          # e.g., "/webshop/articles/4?s=1"
+          def full_path(rack_request)
+            rack_request.fullpath
+          end
+
+          # catch exceptions that may be raised in the middleware chain
+          # Note: if a middleware catches an Exception without re raising,
+          # the Exception cannot be recorded here.
+          def record_and_reraise_error(error, request_span:)
+            request_span.status = OpenTelemetry::Trace::Status.new(
+              OpenTelemetry::Trace::Status::INTERNAL_ERROR,
+              description: error.to_s
+            )
+
+            # TODO: implement span.set_error? (this is a specification-level issue):
+            # request_span.set_error(error) unless request_span.nil?
+            #
+            raise error
+          end
+
+          def set_attributes_after_request(span, status, headers, _response)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+
+            # NOTE: if data is available, it would be good to do this:
+            # set_attribute('http.route', ...
+            # e.g., "/users/:userID?
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            allowed_response_headers(headers).each { |k, v| span.set_attribute(k, v) }
+          end
+
+          # @return Hash
+          def allowed_request_headers(env)
+            {}.tap do |result|
+              self.class.allowed_rack_request_headers.each do |hash|
+                result[self.class.build_attribute_name('http.request.headers.', hash[:header])] = env[hash[:rack_header]] if env.key?(hash[:rack_header])
+              end
+            end
+          end
```

Comment:
I think this optimization has gone awry. Can't we eagerly convert the allowed request headers into the rack normalized versions to use a keys in a map index the attribute names E.g `{'HTTP_FOO' => 'http.request.headers.foo', 'HTTP_BAR' => 'http.request.headers.bar'}`? You can then just iterate through the hash to looking for the key and you'll already have a pre-computed attribute name for it if it's found?